### ISSUE TITLE
Large performance issue in find_bins

### ISSUE THREAD

**Issue by p-slash** [igmhub/picca#978]
Hi all,

I've been sitting on this for a few months, thinking I'll have time to come back to it. At this point it's best if I describe flag the issue here.

Problem: `find_bins` function performs a linear search to find bin indices for rebining, stacking etc. This is an algorithmically slow, O(N), operation. njit helps, but a better method can yield order of magnitude speed up.

Solution: I propose two solutions and leave you to decide which one fits the design best.
1. Fastest algorithm: Binning indices for an *equally* and *linearly* spaced output array can be directly computed by a simple mathematical operation. Say _y_ is the input value, _x0_ is the first edge of the output array and _dx_ is the spacing of the output array. Then, the bin index _idx_ = int( (_y_ - _x0_) / _dx_ ).
2. Binary search: If the method above is not possible, I recommend using `numpy.searchsorted`.

I'd be happy to discuss this in the next telecon @iprafols , @Waelthus .

Best,
Naim

**Comment by iprafols** [comment]
Hi @p-slash,

I've been wanting to improve this function for a while too. I guess we will need to take into account the different ways this function is/can be used (e.g. log binning and linear binning). We could break something if we're not careful (for example if the output array is not equally and linearly spaced. So sure let's discuss this on the next telecon.

**Comment by Waelthus** [comment]
If `find_bins` is currently still a bottleneck in the computation, this is a good idea to speed things up. But note that the current approach finds the closest bin, i.e. sometimes the one that is smaller than original_array[index1], sometimes the one that is larger (which might be an issue on it's own and led to some amount of trouble ~1yr back). 

If we're actually only using that for uniformly linearly spaced `grid_array` then indeed we indeed should do 1. which is an O(1) operation (potentially with a shift of 0.5 to mimic the current behaviour). If those properties are not fulfilled (I seem to remember that the idea behind having this function originally was that we needed something which works on non-uniformly spaced grids as well), then we'd get into trouble replicating the current behaviour with most approaches. In which case we might break backwards-compatibility and need more extensive testing.

**Comment by p-slash** [comment]
What I remember from my tests few months: I implemented `np.searchsorted`. This is easier to add because it supports unequal bin spacing. However, I got stuck at passing the tests. It probably has something to do with what Michael mentioned ` i.e. sometimes the one that is smaller than original_array[index1], sometimes the one that is larger (which might be an issue on it's own and led to some amount of trouble ~1yr back).` If the points exactly falls onto an edge, then it is fine. It is a machine precision problem - not much we can do.
